### PR TITLE
fix(palettization): guarantee calibration cleanup, state recovery, and fix root module sensitivity naming

### LINKED ISSUE #94: [Bug]: KMeansPalettizer leaks checkpoint files and leaves fake palettization disabled on calibration failure; fails to resolve root module sensitivities
### Describe the Bug

Two related issues in `KMeansPalettizer` cause resource leaks, state corruption, and silent sensitivity configuration failures during calibration:

1. **Resource Leak & State Corruption on Calibration Abort (`src/coreai_opt/palettization/kmeans/palettizer.py:285-340`)**:
   When entering `KMeansPalettizer.calibration_mode()`, the palettizer saves a temporary model checkpoint to disk via `tempfile.NamedTemporaryFile(delete=False, suffix=".pt", prefix="palettizer_calibration_")`.
   The deletion `os.unlink(checkpoint_path)` only takes place inside `_load_model_checkpoint()`.
   However, the post-calibration logic was located inside the generator's `finally:` block. Consequently:
   - If an exception is raised inside the user's `with palettizer.calibration_mode(...)` block prior to calling `skm.step()` (e.g. data loader error, OOM during forward pass, or user assertion), the `finally:` block executes `if not calibration_helper.step_called: raise RuntimeError(...)`. In Python, raising in a `finally:` block **completely masks and suppresses the caller's original exception**.
   - Because `_load_model_checkpoint()` is never reached, the multi-gigabyte `.pt` model checkpoint file is **permanently leaked in `/tmp`**. On distributed cluster nodes or shared servers, repeated failed calibration attempts can quickly cause disk exhaustion.
   - `self._model.apply(_disable_fake_palett)` was called at context entry, but `self._model.apply(_enable_fake_palett)` is never reached on abort. The model is left permanently mutated with fake palettization silently disabled.
   - Model weights modified during calibration are left un-restored.

2. **Root Module Sensitivity Parameter Name Mismatch (`src/coreai_opt/palettization/kmeans/palettizer.py:685, 767`)**:
   `_set_sensitivities_in_fake_palettize_modules` and `_get_sensitivities_from_fake_palettize_modules` build parameter names using:
   ```python
   param_name = ".".join([module_name, "parametrizations", attr_name, "original"])
   ```
   When palettizing a standalone root module (e.g., `model = nn.Linear(...)`), `module_name == ""`. This produces `".parametrizations.weight.original"` with an invalid leading dot.
   Because PyTorch's `self._model.named_parameters()` produces `"parametrizations.weight.original"` (without a leading dot), `param_name not in sensitivity_dict` triggers every time, logging:
   ```text
   ERROR: No sensitivity value found for .parametrizations.weight.original
   ```
   Sensitivities are therefore never assigned to root-level modules.

---

#### Case A: Resource leak & exception masking on calibration failure
```python
import glob, os, tempfile
import torch, torch.nn as nn
from coreai_opt.palettization import KMeansPalettizer, KMeansPalettizerConfig, ModuleKMeansPalettizerConfig
from coreai_opt.palettization.spec import default_weight_palettization_spec

model = nn.Sequential(nn.Linear(4, 4))
config = KMeansPalettizerConfig(
    global_config=ModuleKMeansPalettizerConfig(
        op_state_spec={"weight": default_weight_palettization_spec()}
    )
)
palettizer = KMeansPalettizer(model, config)
prepared = palettizer.prepare((torch.randn(2, 4),))

tmp_dir = tempfile.gettempdir()
before = set(glob.glob(os.path.join(tmp_dir, "palettizer_calibration_*.pt")))

class DataLoadingError(Exception): pass

try:
    with palettizer.calibration_mode(loss_fn=lambda out, tgt: out.sum()) as skm:
        raise DataLoadingError("DataLoader failed before step")
except BaseException as e:
    print("Caught:", type(e).__name__, str(e))

after = set(glob.glob(os.path.join(tmp_dir, "palettizer_calibration_*.pt")))
print("Leaked checkpoints count:", len(after - before))
```
**Output on `main`**:
```text
Caught: RuntimeError calibration_mode requires at least one call to step(). No calibration data was processed.
Leaked checkpoints count: 1
```
*(Notice: `DataLoadingError` was masked, and a `.pt` file was leaked).*

#### Case B: Root module sensitivity failure
```python
model = nn.Linear(4, 4)
palettizer = KMeansPalettizer(model, config)
prepared = palettizer.prepare((torch.randn(2, 4),))

with palettizer.calibration_mode(loss_fn=lambda out, tgt: out.sum()) as skm:
    out = prepared(torch.randn(2, 4))
    skm.step(out, torch.zeros(2, 4))
```
**Output on `main`**:
```text
ERROR coreai_opt.palettization.kmeans.palettizer: No sensitivity value found for .parametrizations.weight.original
```

---

### Expected Behavior
1. If an error occurs inside `calibration_mode`:
   - The caller's original exception should propagate without being masked.
   - Temporary `.pt` checkpoint files in `/tmp` must be deterministically unlinked.
   - Model weights must be rolled back to their pre-calibration state.
   - Fake palettization must be re-enabled.
   - Palettizer lifecycle must reset to `IDLE`.
2. Root modules must resolve parameter names canonically (`"parametrizations.weight.original"`) without leading dots, correctly attaching sensitivity tensors.

---

### Environment
- OS: macOS (Darwin)
- Python: 3.11
- PyTorch: 2.13.0
- Commit: `main` (73fd849)

### PR DESCRIPTION
**Pull request by rohith500**
### Summary of Changes

Closes #94.

This PR resolves two production issues in `KMeansPalettizer`:

1. **Deterministic Checkpoint Cleanup & State Rollback in `calibration_mode`**:
   - Refactored `calibration_mode` context manager to cleanly separate the normal commit path from the error rollback path (`finally:`).
   - Ensured that if an exception is raised inside `with palettizer.calibration_mode(...)`:
     - The caller's original exception propagates with its full traceback rather than being masked by a secondary `RuntimeError`.
     - The temporary `.pt` model checkpoint file is guaranteed to be unlinked, preventing disk leaks in `/tmp`.
     - Model weights are rolled back from the checkpoint to their pre-calibration state.
     - Fake palettization is re-enabled (`self._model.apply(_enable_fake_palett)`), ensuring the model is not left in a disabled state.
     - Compressor lifecycle resets to `IDLE`.
   - In `_load_model_checkpoint()`, wrapped `os.unlink` in `try ... finally` to guarantee cleanup even if `torch.load` / `load_state_dict` encounters deserialization errors.

2. **Root Module Parameter Name Normalization**:
   - In `_set_sensitivities_in_fake_palettize_modules` and `_get_sensitivities_from_fake_palettize_modules`, updated parameter name construction to:
     ```python
     param_name = ".".join(filter(None, [module_name, "parametrizations", attr_name, "original"]))
     ```
     This eliminates the invalid leading dot when `module_name == ""` (root module), matching PyTorch's `named_parameters()` key hierarchy (`"parametrizations.weight.original"`).

---

### Motivation & Impact

In large-scale or multi-job model optimization pipelines:
* Model checkpoints saved during calibration can range from hundreds of megabytes to tens of gigabytes. Any aborted run on `main` orphaned these files in `/tmp`, leading to disk space exhaustion.
* Masking caller exceptions hindered debugging by suppressing the root cause traceback.
* Standalone root modules (e.g. single linear/embedding blocks being optimized) failed to receive sensitivities due to leading-dot key mismatch.

---

### Automated Tests Added

Added to `tests/palettization/test_kmeans_palettizer.py`:
* `test_calibration_mode_exception_cleans_up_checkpoint_and_restores_state`:
  - Validates pre-step abort: caller exception is unmasked, temp checkpoint is unlinked, fake palettization is re-enabled, and lifecycle returns to `IDLE`.
  - Validates post-step abort: mutated weights are rolled back to pre-calibration state, temp checkpoint is unlinked, fake palettization is re-enabled, and lifecycle returns to `IDLE`.
  - Uses a process-safe spy pattern on `_save_model_checkpoint` to prevent race conditions during parallel test execution (`pytest -n auto`).
* `test_calibration_mode_root_module`:
  - Validates that a standalone root `nn.Linear` module completes calibration and successfully assigns normalized sensitivity buffers to `_KMeansFakePalettize`.

---

### Verification

* `pytest tests/palettization/test_kmeans_palettizer.py`: **83 passed in 7.59s**
* `pytest tests/palettization/ -m "not slow" -n auto`: **420 passed in 14.30s**
* `make check`: **All checks passed** (`ruff`, `darker`, license header, towncrier check, etc.)

---

### Checklist
- [x] Follows Conventional Commits formatting
- [x] Commit is cryptographically signed with SSH key
- [x] Includes Towncrier fragment (`changelog.d/94.fixed`)
- [x] Passes `make check` without warnings or diff pollution
- [x] Includes automated tests reproducing the bug and proving the fix

### CODE REVIEW

**Review by chaichanachai54-bit — COMMENTED** [review]
Apple

**Review by chaichanachai54-bit — COMMENTED** [review]
Apple

**Review by chaichanachai54-bit — COMMENTED** [review]
Apple

**Review by chaichanachai54-bit — COMMENTED** [review]
Apple

File: src/coreai_opt/palettization/kmeans/palettizer.py [hunk 1]
```diff
@@ -305,38 +306,48 @@ def step(self, output: torch.Tensor, target: torch.Tensor):
...
+            # parallelism the user opted into at prepare() time.
+            if self._num_workers > 1:
+                self._calculate_centroids_parallel(self._num_workers)
+            else:
+                self._calculate_centroids_sequential()
 
-                    # Restore normal operation
-                    self._model.apply(_enable_fake_palett)
         finally:
+            if checkpoint_path is not None and os.path.exists(checkpoint_path):
```
Comment (crowbat) [inline src/coreai_opt/palettization/kmeans/palettizer.py]: 2 comments here:
1) We should also include self._model.zero_grad() here as part of model restoration (and we can remove it from calibration_mode's try block).

2) I'm thinking that if the loading of the checkpoint fails, that should be a loud error to the user that the model is not in a good state, instead of silently passing any exceptions through. What do you think of this reframing (all living inside the finally block):

```
self._model.zero_grad()
self._model.apply(_enable_fake_palett)
self._lifecycle = _CompressorLifecycle.IDLE
if checkpoint_path is not None:
    try:
        self._load_model_checkpoint(self._model, checkpoint_path)
    except Exception as e:
        raise RuntimeError(
            "Failed to restore the model to its pre-calibration state after "
            "calibration_mode aborted. The model is now in an inconsistent "
            "state and must not be used — reload it before retrying."
        ) from e
```

Removing `os.path.exists(checkpoint_path)` from the if check makes things more robust since if `checkpoint_path` is not None but the path doesn't exist for some reason, then things are really messed up and we could let `_load_model_checkpoint`'s internal if check raise its exception too.

**Review by crowbat — COMMENTED** [review]
Hi @rohith500 , thanks a lot for your cleanup and bugfix. I just have two comments surrounding the finally block in `calibration_mode()`.

File: src/coreai_opt/palettization/kmeans/palettizer.py (hunk 1 above)
Comment (rohith500) [reply src/coreai_opt/palettization/kmeans/palettizer.py]: Hi @crowbat, thanks for the review and great suggestions!

I have addressed both items in commit a105d92:
1. Moved `self._model.zero_grad()` into `finally:` to ensure residual gradients are always cleaned up regardless of whether calibration succeeds or aborts, and removed the duplicate call from the `try:` block.
2. Adopted your reframing for the checkpoint rollback in `finally:`. If `self._load_model_checkpoint` fails during rollback, it now loudly raises a `RuntimeError` notifying the user of the inconsistent state.
3. Added regression tests verifying gradient zeroing on abort and the descriptive `RuntimeError` if checkpoint restoration fails.

All checks (`make check`) and the full palettization test suite (421 tests) are passing.

**Review by crowbat — APPROVED** [review]
Thanks for the fixes @rohith500 , looks good!
